File: backend/app/models/analysis_history.py
"""Persistent analysis history model."""

from datetime import datetime

from sqlalchemy import DateTime, Float, Integer, String, Text
from sqlalchemy.dialects.postgresql import JSONB
from sqlalchemy.orm import Mapped, mapped_column

from app.database.database import Base


class AnalysisHistory(Base):
    __tablename__ = "analysis_history"

    id: Mapped[int] = mapped_column(
        Integer,
        primary_key=True,
        index=True,
    )

    analysis_id: Mapped[str] = mapped_column(
        String(32),
        unique=True,
        nullable=False,
        index=True,
    )

    client_id: Mapped[str] = mapped_column(
        String(128),
        nullable=False,
        index=True,
    )

    latitude: Mapped[float] = mapped_column(
        Float,
        nullable=False,
    )

    longitude: Mapped[float] = mapped_column(
        Float,
        nullable=False,
    )

    location_name: Mapped[str | None] = mapped_column(
        String(255),
        nullable=True,
    )

    project_type: Mapped[str] = mapped_column(
        String(32),
        nullable=False,
    )

    installation_type: Mapped[str] = mapped_column(
        String(32),
        nullable=False,
    )

    land_area_hectares: Mapped[float] = mapped_column(
        Float,
        nullable=False,
    )

    available_budget: Mapped[float] = mapped_column(
        Float,
        nullable=False,
    )

    overall_site_score: Mapped[float] = mapped_column(
        Float,
        nullable=False,
    )

    recommended_deployment: Mapped[str] = mapped_column(
        String(64),
        nullable=False,
    )

    status: Mapped[str] = mapped_column(
        String(32),
        nullable=False,
        default="Completed",
    )

    request_data: Mapped[dict] = mapped_column(
        JSONB,
        nullable=False,
    )

    response_data: Mapped[dict] = mapped_column(
        JSONB,
        nullable=False,
    )

    created_at: Mapped[datetime] = mapped_column(
        DateTime,
        default=datetime.utcnow,
        nullable=False,
        index=True,
    )

File: frontend/src/lib/api/analysis-history.ts
import { apiDownload, apiDownloadFile, apiRequest } from "./client";

export type AnalysisHistoryApiRecord = {
  id: number;
  analysis_id: string;
  location_name: string | null;
  latitude: number;
  longitude: number;
  project_type: "solar" | "wind" | "hybrid" | string;
  installation_type: "ground-mounted" | "rooftop" | "other" | string;
  land_area_hectares: number;
  available_budget: number;
  overall_site_score: number;
  recommended_deployment: string;
  status: "Completed" | "Failed" | "Running" | string;
  created_at: string;
  response_data: Record<string, unknown>;
};

export type AnalysisHistoryListResponse = {
  items: AnalysisHistoryApiRecord[];
  total: number;
  page: number;
  page_size: number;
  pages: number;
};

export type AnalysisHistoryDetail = AnalysisHistoryApiRecord & {
  request_data: Record<string, unknown>;
};

export function listAnalysisHistory(
  page = 1,
  pageSize = 100,
  query?: string,
): Promise<AnalysisHistoryListResponse> {
  const params = new URLSearchParams({
    page: String(page),
    page_size: String(pageSize),
  });

  if (query?.trim()) {
    params.set("query", query.trim());
  }

  return apiRequest<AnalysisHistoryListResponse>(`/analysis-history?${params.toString()}`);
}

export async function listAllAnalysisHistory(): Promise<AnalysisHistoryApiRecord[]> {
  const firstPage = await listAnalysisHistory(1, 100);
  const remainingPages = await Promise.all(
    Array.from({ length: Math.max(firstPage.pages - 1, 0) }, (_, index) =>
      listAnalysisHistory(index + 2, 100),
    ),
  );

  return [firstPage, ...remainingPages].flatMap((page) => page.items);
}

export function getAnalysisHistory(analysisId: string): Promise<AnalysisHistoryDetail> {
  return apiRequest<AnalysisHistoryDetail>(`/analysis-history/${encodeURIComponent(analysisId)}`);
}

export function deleteAnalysisHistory(
  analysisId: string,
): Promise<{ status: string; analysis_id: string }> {
  return apiRequest<{ status: string; analysis_id: string }>(
    `/analysis-history/${encodeURIComponent(analysisId)}`,
    { method: "DELETE" },
  );
}

export function downloadAnalysisHistory(analysisId: string): Promise<Blob> {
  return apiDownload(`/analysis-history/${encodeURIComponent(analysisId)}/download`);
}

export function downloadAnalysisHistoryFile(
  analysisId: string,
): Promise<{ blob: Blob; filename: string }> {
  return apiDownloadFile(
    `/analysis-history/${encodeURIComponent(analysisId)}/download`,
    `${analysisId}.pdf`,
  );
}
